drivers/led_strip/tlc59731.c

```diff
@@ -0,0 +1,190 @@
+ * is needed there.
+ * A zero is represented by no additional pulses within a cycle.
+ * A one is represented by an additional pulse between 275ns and 2.5us
+ * (half a cycle) after the first one. We need at least some delay to get to
+ * 275ns, but because of the limited granularity of k_busy_wait we use a
+ * full 1us. After the pulse, we wait an additional T_CYCLE_1 to complete
+ * the cycle. This time can be slightly shorter because the second pulse
+ * already closes the cycle.
+ * Finally we need to keep the line low for T_H0 to complete the address
+ * for a single chip, and T_H1 to complete the write for all chips.
+ */
+
+#include <zephyr/drivers/led_strip.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(tlc59731, CONFIG_LED_STRIP_LOG_LEVEL);
+
+/* Pulse timing */
+#define TLC59731_DELAY     0x01 /* us */
+#define TLC59731_T_CYCLE_0 0x04 /* us */
+#define TLC59731_T_CYCLE_1 0x01 /* us */
+#define TLC59731_T_H0      (4 * TLC59731_T_CYCLE_0)
+#define TLC59731_T_H1      (8 * TLC59731_T_CYCLE_0)
+/* Threshould levels */
+#define TLC59731_HIGH      0x01
+#define TLC59731_LOW       0x00
+
+/* Write command */
+#define TLC59731_WR 0x3A
+
+struct tlc59731_cfg {
+	struct gpio_dt_spec sdi_gpio;
+};
+
+static inline int rgb_pulse(const struct gpio_dt_spec *led_dev)
+{
+	int fret = 0;
+
+	fret = gpio_pin_set_dt(led_dev, TLC59731_HIGH);
+	if (fret != 0) {
+		return fret;
+	}
+
+	fret = gpio_pin_set_dt(led_dev, TLC59731_LOW);
+	if (fret != 0) {
+		return fret;
+	}
+
+	return fret;
+}
+
+static int rgb_write_bit(const struct gpio_dt_spec *led_dev, uint8_t data)
+{
+	rgb_pulse(led_dev);
+
+	k_busy_wait(TLC59731_DELAY);
+
+	if (data) {
+		rgb_pulse(led_dev);
+		k_busy_wait(TLC59731_T_CYCLE_1);
+	} else {
+		k_busy_wait(TLC59731_T_CYCLE_0);
+	}
+
+	return 0;
+}
+
+static int rgb_write_data(const struct gpio_dt_spec *led_dev, uint8_t data)
+{
+	int8_t idx = 7;
+
+	while (idx >= 0) {
+		rgb_write_bit(led_dev, data & BIT((idx--)));
+	}
+
+	return 0;
+}
+
+static int tlc59731_led_set_color(const struct device *dev, struct led_rgb *pixel)
+{
+
+	const struct tlc59731_cfg *tlc_conf = dev->config;
+	const struct gpio_dt_spec *led_gpio = &tlc_conf->sdi_gpio;
+
+	rgb_write_data(led_gpio, TLC59731_WR);
+	rgb_write_data(led_gpio, pixel->r);
+	rgb_write_data(led_gpio, pixel->g);
+	rgb_write_data(led_gpio, pixel->b);
+
+	return 0;
+}
+
+static int tlc59731_gpio_update_rgb(const struct device *dev, struct led_rgb *pixels,
+				    size_t num_pixels)
+{
+	size_t i;
+	int err = 0;
+
+	for (i = 0; i < num_pixels; i++) {
+		err = tlc59731_led_set_color(dev, &pixels[i]);
+		if (err) {
+			break;
+		}
+	}
+
+	return err;
+}
+
+static int tlc59731_gpio_update_channels(const struct device *dev, uint8_t *channels,
+					 size_t num_channels)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(channels);
+	ARG_UNUSED(num_channels);
+
+	return -ENOTSUP;
```

Comment:
@simonguinot and also for this to error if not supported in the same include so it can be entirely omitted here

Reply:
Yes

Reply:
This function can go